// scratch-vm/src/extensions/scratch3_dataProcessing/index.js
const ArgumentType = require('../../extension-support/argument-type');
const BlockType = require('../../extension-support/block-type');
const msg = require('./translation');
const formatMessage = require('format-message');

const menuIconURI = null;
const blockIconURI = null;

const defaultId = 'default';
let theLocale = null;

class dataProcessing {
    constructor(runtime) {
        theLocale = this._setLocale();
        this.runtime = runtime;
        // communication related
        this.comm = runtime.ioDevices.comm;
        this.session = null;
        this.runtime.registerPeripheralExtension('chart', this);
        // session callbacks
        this.reporter = null;
        this.onmessage = this.onmessage.bind(this);
        this.onclose = this.onclose.bind(this);
        this.write = this.write.bind(this);

    }

    onclose() {
        this.session = null;
    }

    write(data, parser = null) {
        if (this.session) {
            return new Promise(resolve => {
                if (parser) {
                    this.reporter = {
                        parser,
                        resolve
                    };
                }
                this.session.write(data);
            });
        }
    }

    onmessage(data) {
        const dataStr = this.decoder.decode(data);
        this.lineBuffer += dataStr;
        if (this.lineBuffer.indexOf('\n') !== -1) {
            const lines = this.lineBuffer.split('\n');
            this.lineBuffer = lines.pop();
            for (const l of lines) {
                if (this.reporter) {
                    const { parser, resolve } = this.reporter;
                    resolve(parser(l));
                }
            }
        }
    }

    scan() {
        this.comm.getDeviceList().then(result => {
            this.runtime.emit(this.runtime.constructor.PERIPHERAL_LIST_UPDATE, result);
        });
    }

    _setLocale() {
        let nowLocale = '';
        switch (formatMessage.setup().locale) {
            case 'zh-tw':
                nowLocale = 'zh-tw';
                break;
            default:
                nowLocale = 'en';
                break;
        }
        return nowLocale;
    }

    getInfo() {
        theLocale = this._setLocale();

        return {
            id: 'dataProcessing',
            name: msg.title[theLocale],
            color1: '#566988',
            color2: '#566988',
            menuIconURI: menuIconURI,
            blockIconURI: blockIconURI,
            blocks: [
                {
                    opcode: 'conversion',
                    blockType: BlockType.REPORTER,
                    arguments: {
                        before: {
                            type: ArgumentType.STRING,
                            menu: 'selectYear',
                            defaultValue: '二進制'
                        },
                        data: {
                            type: ArgumentType.STRING,
                            defaultValue: '1010'
                        },
                        after: {
                            type: ArgumentType.STRING,
                            menu: 'selectYear',
                            defaultValue: '十進制'
                        },
                    },
                    text: msg.conversion[theLocale]
                },
                '---',
                {
                    opcode: 'extractText',
                    blockType: BlockType.REPORTER,
                    arguments: {
                        data: {
                            type: ArgumentType.STRING,
                            defaultValue: 'data'
                        },
                        start: {
                            type: ArgumentType.STRING,
                            defaultValue: '1'
                        },
                        finish: {
                            type: ArgumentType.STRING,
                            defaultValue: '1'
                        },
                    },
                    text: msg.extractText[theLocale]
                },
                {
                    opcode: 'textsplit',
                    blockType: BlockType.REPORTER,
                    arguments: {
                        TEXT: {
                            type: ArgumentType.STRING,
                            defaultValue: 'abc,def'
                        },
                        SPLIT: {
                            type: ArgumentType.STRING,
                            defaultValue: ','
                        }
                    },
                    text: msg.textsplit[theLocale]
                },
            ],
            menus: {
                selectYear: {
                    acceptReporters: true,
                    items: ["二進制", "八進制", "十進制", "十六進制"],
                },
            }
        };
    }

    conversion(args) {
        var beforeConversion = args.before;
        var afterConversion = args.after;
        var data = args.data;

        var convertToDecimal;
        var convertToOtherBases;

        switch (beforeConversion) {
            case "二進制": convertToDecimal = parseInt(data, 2); break;
            case "八進制": convertToDecimal = parseInt(data, 8); break;
            case "十六進制": convertToDecimal = parseInt(data, 16); break;
            default: convertToDecimal = parseInt(data, 10); break;
        }

        switch (afterConversion) {
            case "二進制": convertToOtherBases = convertToDecimal.toString(2) ; break;
            case "八進制": convertToOtherBases = convertToDecimal.toString(8); break;
            case "十六進制": convertToOtherBases = convertToDecimal.toString(16); break;
            default: convertToOtherBases = convertToDecimal.toString(10);; break;
        }

        return convertToOtherBases.toString();
    }

    extractText(args) {
        var data = args.data;
        var start = args.start;
        var finish = args.finish;

        var text = data.substring(start - 1, finish);

        return text.toString();
    }

    textsplit(args){
        const args_text = args.TEXT;
        const split_text = args.SPLIT;
        const sp_ary = args_text.split(split_text);
        const r_json = JSON.stringify(Object.assign({}, sp_ary));        
        console.log(r_json);
        return r_json;

    }

}

module.exports = dataProcessing;
