// scratch-vm/src/extensions/scratch3_json/index.js
const ArgumentType = require('../../extension-support/argument-type');
const BlockType = require('../../extension-support/block-type');
const msg = require('./translation');
const formatMessage = require('format-message');
const xml2js = require('xml2js');
//async add estea
const ml5 = require('ml5');
//require('babel-polyfill');
//end
const menuIconURI = null;
const blockIconURI = null;
//2.4
let cors_url ='https://script.google.com/macros/s/AKfycbzF0Bvv27nBC3xXdAYRzM4ZIvoipG0cOHVM_NniomcySjs0PKyQHBBKJBje9q-oeZTW/exec';

const defaultId = 'default';
let theLocale = null;

class gasoJSON {
    constructor(runtime) {
        theLocale = this._setLocale();
        this.runtime = runtime;
        // communication related
        this.comm = runtime.ioDevices.comm;
        this.session = null;
        this.runtime.registerPeripheralExtension('gasoJSON', this);
        // session callbacks
        this.reporter = null;
        this.onmessage = this.onmessage.bind(this);
        this.onclose = this.onclose.bind(this);
        this.write = this.write.bind(this);
        // string op
        this.decoder = new TextDecoder();
        this.lineBuffer = '';
        this.data = {};
        //estea
        this.jsoncol = {};
        //estea end
        this.emptyObj = {
            VALUE: {}
        };
    }

    onclose() {
        this.session = null;
    }

    write(data, parser = null) {
        if (this.session) {
            return new Promise(resolve => {
                if (parser) {
                    this.reporter = {
                        parser,
                        resolve
                    };
                }
                this.session.write(data);
            });
        }
    }

    onmessage(data) {
        const dataStr = this.decoder.decode(data);
        this.lineBuffer += dataStr;
        if (this.lineBuffer.indexOf('\n') !== -1) {
            const lines = this.lineBuffer.split('\n');
            this.lineBuffer = lines.pop();
            for (const l of lines) {
                if (this.reporter) {
                    const { parser, resolve } = this.reporter;
                    resolve(parser(l));
                }
            }
        }
    }

    scan() {
        this.comm.getDeviceList().then(result => {
            this.runtime.emit(this.runtime.constructor.PERIPHERAL_LIST_UPDATE, result);
        });
    }

    _setLocale() {
        let nowLocale = '';
        switch (formatMessage.setup().locale) {
            case 'zh-tw':
                nowLocale = 'zh-tw';
                break;
            default:
                nowLocale = 'en';
                break;
        }
        return nowLocale;
    }

    getInfo() {
        theLocale = this._setLocale();

        return {
            id: 'gasoJSON',
            name: 'JSON',
            color1: '#4a90e2',
            color2: '#4a90e2',
            menuIconURI: menuIconURI,
            blockIconURI: blockIconURI,
            blocks: [
                {
                    opcode: 'fetchJSON',
                    blockType: BlockType.COMMAND,
                    arguments: {
                        url: {
                            type: ArgumentType.STRING,
                            defaultValue: 'https://'
                        },
                    },
                    text: msg.fetchJSON[theLocale]
                },
                {
                    opcode: 'fetchCSV',
                    blockType: BlockType.COMMAND,
                    arguments: {
                        url: {
                            type: ArgumentType.STRING,
                            defaultValue: 'https://'
                        },
                    },
                    text: msg.fetchCSV[theLocale]
                },
                {
                    opcode: 'fetchTXT',
                    blockType: BlockType.COMMAND,
                    arguments: {
                        url: {
                            type: ArgumentType.STRING,
                            defaultValue: 'https://'
                        },
                    },
                    text: msg.fetchTXT[theLocale]
                },
                {
                    opcode: 'fetchXML',
                    blockType: BlockType.COMMAND,
                    arguments: {
                        url: {
                            type: ArgumentType.STRING,
                            defaultValue: 'https://'
                        },
                    },
                    text: msg.fetchXML[theLocale]
                },
                {
                    opcode: 'openCorsSite',
                    blockType: BlockType.COMMAND,
                    text: msg.openCORS[theLocale]
                },
                {
                    opcode: "clearJSONData",
                    blockType: BlockType.COMMAND,
                    text: msg.clearJSONData[theLocale],
                },
                "---",
                {
                    opcode: 'onJSONReceived',
                    blockType: BlockType.HAT,
                    isEdgeActivated: false,
                    arguments: {},
                    text: msg.onJSONReceived[theLocale]
                },
                "---",
                {
                    opcode: 'readFromJSON',
                    blockType: BlockType.REPORTER,
                    arguments: {
                        id: {
                            type: ArgumentType.STRING,
                            defaultValue: 'id'
                        }
                    },
                    text: msg.readFromJSON[theLocale]
                },
                {
                    opcode: 'readEntryFromJSON',
                    blockType: BlockType.REPORTER,
                    arguments: {
                        variable: {
                            type: ArgumentType.STRING,
                            defaultValue: msg.data[theLocale]
                        },
                        n: {
                            type: ArgumentType.NUMBER,
                            defaultValue: '1'
                        }
                    },
                    text: msg.readEntryFromJSON[theLocale]
                },
                {
                    opcode: 'readAttrFromJSON',
                    blockType: BlockType.REPORTER,
                    arguments: {
                        variable: {
                            type: ArgumentType.STRING,
                            defaultValue: msg.data[theLocale]
                        },
                        attr: {
                            type: ArgumentType.STRING,
                            defaultValue: msg.attr[theLocale]
                        }
                    },
                    text: msg.readAttrFromJSON[theLocale]
                },
                {
                    opcode: 'getJsonCol',
                    blockType: BlockType.REPORTER,
                    arguments: {
                        variable: {
                            type: ArgumentType.STRING,
                            defaultValue: msg.data[theLocale]
                        },
                    },
                    text: msg.getJsonCol[theLocale]
                },
                {
                    opcode: 'getJsonAllNum',
                    blockType: BlockType.REPORTER,
                    arguments: {
                        variable: {
                            type: ArgumentType.STRING,
                            defaultValue: msg.data[theLocale]
                        }
                    },
                    text: msg.jsonFormNum[theLocale]
                },
                "---",
                {
                    opcode: "useTextSearchGetData",
                    blockType: BlockType.REPORTER,
                    arguments: {
                        data: {
                            type: ArgumentType.STRING,
                            defaultValue: msg.data[theLocale],
                        },
                        column: {
                            type: ArgumentType.STRING,
                            defaultValue: msg.Column_name[theLocale],
                        },
                        value: {
                            type: ArgumentType.STRING,
                            defaultValue: msg.value[theLocale],
                        },
                        column2: {
                            type: ArgumentType.STRING,
                            defaultValue: msg.Column_name[theLocale],
                        },
                    },
                    text: msg.useTextSearchGetData[theLocale],
                },
                {
                    opcode: "useNumberCompareGetData",
                    blockType: BlockType.REPORTER,
                    arguments: {
                        data: {
                            type: ArgumentType.STRING,
                            defaultValue: msg.data[theLocale],
                        },
                        column: {
                            type: ArgumentType.STRING,
                            defaultValue: msg.Column_name[theLocale],
                        },
                        symbol: {
                            type: ArgumentType.STRING,
                            defaultValue: ">",
                            menu: "symbolItem",
                        },
                        number: {
                            type: ArgumentType.NUMBER,
                            defaultValue: msg.number[theLocale],
                        },
                        column2: {
                            type: ArgumentType.STRING,
                            defaultValue: msg.Column_name[theLocale],
                        },
                    },
                    text: msg.useNumberCompareGetData[theLocale],
                },
                {
                    opcode: "splitString",
                    blockType: BlockType.REPORTER,
                    arguments: {
                        data: {
                            type: ArgumentType.STRING,
                            defaultValue: msg.data[theLocale],
                        },
                        n: {
                            type: ArgumentType.NUMBER,
                            defaultValue: "1",
                        },
                    },
                    text: msg.splitString[theLocale],
                },
            ],
            menus: {
                symbolItem: {
                    acceptReporters: true,
                    items: [">", ">=", "!=", "=", "<=", "<"],
                },
                file_type: {
                    acceptReporters: true,
                    items: ["json", "txt", "csv"],
                }
            }
        };
    }

    async fetchJSON(args) {
        const url = encodeURIComponent(args.url);
        const f_type = 'json';
        console.log(url);
        //let cors_url = 'https://script.google.com/macros/s/AKfycbzE3fs6P0HAkK0dUE31wLambBLhjtsRK1zY3ZytotQenoVNpaAsunXhz9lqKgoiEpp9/exec';
        console.log(cors_url +'?url='+ url);
        return await fetch(cors_url +'?f='+f_type+'&url='+ url).then(res => {
           
            if (res.ok) {
                res.json().then(json => {
                    this.data.fetched = true;
                    this.data.data = JSON.stringify(json);
                    this.runtime.startHats('gasoJSON_onJSONReceived', {});
                });
            }
        })
        
    }

    async fetchCSV(args) {
        const url = encodeURIComponent(args.url);
        const f_type = 'csv';
        console.log(cors_url +'?url='+ url);
        return await fetch(cors_url +'?f='+f_type+'&url='+ url).then(res => {
           
            if (res.ok) {
                res.json().then(json => {
                    this.data.fetched = true;
                    this.data.data = JSON.stringify(Object.assign({}, json.split('\r\n')));
                    if(f_type=='txt'){
                        this.data.data = JSON.stringify(Object.assign({}, json.split('\n')));
                    };
                    
                    this.runtime.startHats('gasoJSON_onJSONReceived', {});
                });
            }
        })
    }

    async fetchTXT(args) {
        const url = encodeURIComponent(args.url);
        const f_type = 'txt';
        console.log(cors_url +'?url='+ url);
        return await fetch(cors_url +'?f='+f_type+'&url='+ url).then(res => {
           
            if (res.ok) {
                res.json().then(json => {
                    this.data.fetched = true;
                    this.data.data = JSON.stringify(Object.assign({}, json.split('\n')));
                    this.runtime.startHats('gasoJSON_onJSONReceived', {});
                });
            }
        })
    }

    async fetchXML(args) {
        const url = encodeURIComponent(args.url);
        const f_type = 'xml';
        console.log(cors_url +'?f='+f_type+'&url='+ url);
        return await fetch(cors_url +'?f='+f_type+'&url='+ url).then(res => {
            if (res.ok) {
                res.json().then(json => {
                    this.data.fetched = true;
                    xml2js.parseString(json, (err, result) => {
                        if (err) {
                          throw err
                        }
                        // `result` is a JavaScript object
                        // convert it to a JSON string
                        this.data.data = JSON.stringify(result, null, 4)
                        // log JSON string
                        //console.log(this.data.data)
                      })
                    this.runtime.startHats('gasoJSON_onJSONReceived', {});
                });
            }
        })
    }
    
      
    openCorsSite(args) {
        window.open('https://cors-anywhere.herokuapp.com/corsdemo');
    }

    clearJSONData() {
        return (this.data = {});
    }

    isDataFetched() {
        return this.data.fetched;
    }

    onJSONReceived() {
        if (this.isDataFetched()) {
            console.log('got data');
            return true;
        }
    }

    readFromJSON() {
        if (this.isDataFetched()) {
            console.log('return ', this.data.data);
            return this.data.data;
        }
        return msg.readFromJSONErr[theLocale];
    }

    readEntryFromJSON(args) {
        const variable = args.variable || this.emptyObj;
        const n = args.n;
        try {
            const parsed = JSON.parse(variable);
            const data = parsed[n - 1];
            return typeof data === 'string' ? data : JSON.stringify(data);
        } catch (err) {
            return `Error: ${err}`;
        }
    }

    readAttrFromJSON(args) {
        const variable = args.variable || this.emptyObj;
        const attr = args.attr;
        try {
            const parsed = JSON.parse(variable);
            const data = parsed[attr];
            return typeof data === 'string' ? data : JSON.stringify(data);
        } catch (err) {
            return `Error: ${err}`;
        }
    }

    getJsonCol(args) {
        //data 
        const variable = args.variable || this.emptyObj;
        let parsed = JSON.parse(variable);
        //const first_index = parsed[0];
        const first_data = typeof parsed === 'string' ? data : JSON.stringify(parsed);
        console.log(first_data);
        this.jsoncol.fetched = true;
        this.jsoncol.data = Object.keys(JSON.parse(first_data));
        console.log(this.jsoncol.data);
        return JSON.stringify(this.jsoncol.data);
    }

    getJsonAllNum(args) {
        const variable = args.variable || this.emptyObj;
        const parsed = JSON.parse(variable);
        const size = parsed.length;
        console.log(size);
        return size;
    }

    useTextSearchGetData(args) {
        const data = args.data;
        const column = args.column;
        const column2 = args.column2;
        const value = args.value;

        try {
            const parsed = JSON.parse(data);
            const size = parsed.length;
            var getData = "";
            for (var i = 0; i < size; i++) {
                const d = parsed[i][column];
                if (d.includes(value)) getData += parsed[i][column2] + ",";
            }
            return getData.slice(0, getData.length - 1);
        } catch (err) {
            return `Error: ${err}`;
        }
    }

    useNumberCompareGetData(args) {
        const data = args.data;
        const column = args.column;
        const symbol = args.symbol;
        const number = parseInt(args.number, 10);
        const column2 = args.column2;

        try {
            const parsed = JSON.parse(data);
            const size = parsed.length;
            var getData = "";

            for (var i = 0; i < size; i++) {
                const d = parsed[i][column];

                if (symbol == ">" && d > number)
                    getData += parsed[i][column2] + ",";
                if (symbol == ">=" && d >= number)
                    getData += parsed[i][column2] + ",";
                if (symbol == "!=" && d != number)
                    getData += parsed[i][column2] + ",";
                if (symbol == "=" && d == number)
                    getData += parsed[i][column2] + ",";
                if (symbol == "<=" && d <= number)
                    getData += parsed[i][column2] + ",";
                if (symbol == "<" && d < number)
                    getData += parsed[i][column2] + ",";
            }
            return getData.slice(0, getData.length - 1);
        } catch (err) {
            return `Error: ${err}`;
        }
    }

    splitString(args) {
        const data = args.data;
        const n = args.n;
        var splitText = data.split(",");

        try {
            return splitText[n - 1];
        } catch (err) {
            return `Error: ${err}`;
        }
    }
}

module.exports = gasoJSON;
